### PR TITLE
Data import

### PR DESCRIPTION
**Pull request by Rand3rino**
A couple things I am aware of:
CSV nulls were manually turned into " ", I'll look into code that automatically formats this
Exception checking when there's mismatched datatypes is not implemented yet

Can you look this over and see if this is what we need?

### CODE REVIEW

File: Nesteo.Server.DataImport/Program.cs [hunk 1]
```diff
@@ -0,0 +1,420 @@
...
+            {
+                regionEntity = (from r in dbContext.Regions where r.Name == record.Ort select r).FirstOrDefault();
+            }
+            else
+            {
+                regionEntity = null;
+            }
+
+            // Get Material Value
+            if (record.Material.StartsWith("Holbeton"))
```
Comment (MarcusWichelmann) [inline Nesteo.Server.DataImport/Program.cs]: Holzbeton

**Comment by Rand3rino** [comment]
I also noticed that I failed to import all rows, 260/400 Stammdaten and 121/5386 Kontrolldaten. I'm not sure why

File: Nesteo.Server.DataImport/Program.cs [hunk 2]
```diff
@@ -0,0 +1,420 @@
...
+            }
+
+            ReadStammDaten(dbContext, user);
+            await SaveDatabase(dbContext);
+
+            ReadKontrollDaten(dbContext, user);
+            await SaveDatabase(dbContext);
+        }
+
+        private static async Task SaveDatabase(NesteoDbContext dbContext)
```
Comment (MarcusWichelmann) [inline Nesteo.Server.DataImport/Program.cs]: It's convention to name all methods returning a Task like `DoSomethingAsync`. The `Async` helps you not forgetting to add an `await` when calling these methods.

File: Nesteo.Server.DataImport/Program.cs [hunk 3]
```diff
@@ -0,0 +1,420 @@
...
+        {
+            Console.WriteLine("Saving");
+            await dbContext.SaveChangesAsync().ConfigureAwait(false);
+            Console.WriteLine("Done.");
+        }
+
+        private static async void ReadStammDaten(NesteoDbContext dbContext, UserEntity user)
```
Comment (MarcusWichelmann) [inline Nesteo.Server.DataImport/Program.cs]: Haha, it's strange that an american guy uses german method names. I hate our professors at university for doing this, we name this mixture `Denglisch` :smile: 
What about `ReadNestingBoxDataAsync`?

Comment (MarcusWichelmann) [inline Nesteo.Server.DataImport/Program.cs]: Ah, what you probably didn't know yet: `async void` is really bad and could break exception handling when not being very careful. Always use `async Task` please.

File: Nesteo.Server.DataImport/Program.cs [hunk 4]
```diff
@@ -0,0 +1,420 @@
...
+        {
+            using (var reader = new StreamReader("/home/randy/Nesteo-Server/Nesteo.Server.DataImport/Data/Stammdaten-9.csv"))
```
Comment (MarcusWichelmann) [inline Nesteo.Server.DataImport/Program.cs]: Maybe we can make the path selection more dynamic somehow? A start parameter for the application?

File: Nesteo.Server.DataImport/Program.cs [hunk 5]
```diff
@@ -0,0 +1,420 @@
...
+            {
+                material = Material.UntreatedWood;
+            }
+            else if (record.Material == "Holz beschicht")
+            {
+                material = Material.TreatedWood;
+            }
+            else
+            {
+                material = Material.Other;
```
Comment (MarcusWichelmann) [inline Nesteo.Server.DataImport/Program.cs]: Maybe print these out so we can make sure that nothing get's wrongly detected as `unknown`?

File: Nesteo.Server.DataImport/Program.cs [hunk 6]
```diff
@@ -0,0 +1,420 @@
...
+                holeSize = HoleSize.Small;
+            }
+            else if (record.Loch == "Halbhöhle")
+            {
+                // TODO This may be wrong translation
+                holeSize = HoleSize.Oval;
+            }
+            else
+            {
+                holeSize = HoleSize.Other;
```
Comment (MarcusWichelmann) [inline Nesteo.Server.DataImport/Program.cs]: Same here

File: Nesteo.Server.DataImport/Program.cs [hunk 7]
```diff
@@ -0,0 +1,420 @@
...
+                MaleParentBirdDiscovery = ParentBirdDiscovery.None,
+                Species = speciesEntity,
+                ImageFileName = null,
+                Comment = record.Bemerkungen == " "? null: record.Bemerkungen
+
+            };
+            dbContext.Inspections.Add(inspectionEntity);
+        }
+
+        private static T GetRandomOrDefault<T>(this LocalView<T> localView) where T : class => localView.OrderBy(r => Guid.NewGuid()).FirstOrDefault();
```
Comment (MarcusWichelmann) [inline Nesteo.Server.DataImport/Program.cs]: I think you can remove this

**Comment by MarcusWichelmann** [comment]
Also some general things:
- Clean up blank lines
- Move all classes to seperate files so the files don't get too large (Maybe a sub-directory for the CsvEntities..?)
- May be safer to use a case invariant comparison when matching materials/hole sizes/... (`record.Material == "Holz unbeschicht"`)

File: Nesteo.Server.DataImport/Program.cs (hunk 3 above)
Comment (Rand3rino) [reply Nesteo.Server.DataImport/Program.cs]: I'm putting `Denglisch` on my resume 😂

**Comment by Rand3rino** [comment]
I think I addressed the comments you made as well as added the exception handling.

**Comment by Rand3rino** [comment]
@MarcusWichelmann I made changes to the user input and the exception logging. I'm ready for review now

File: Nesteo.Server.DataImport/Program.cs [hunk 8]
```diff
@@ -0,0 +1,443 @@
...
+//            ClearDatabaseAsync(dbContext);
+
+            UserEntity user = await dbContext.Users.FirstOrDefaultAsync().ConfigureAwait(false);
+            if (user == null)
+            {
+                Console.WriteLine("Please run the server first to make sure that at least one user exists.");
+                return;
+            }
+
+            int exceptions = 0;
```
[the remark this answers is not in the document]
Comment (MarcusWichelmann) [reply Nesteo.Server.DataImport/Program.cs]: Oups, haven't seen the console output. Then it's fine.

**Comment by Rand3rino** [comment]
I see your comments and will start resolving these over the weekend.

**Comment by MarcusWichelmann** [comment]
@Rand3rino Please wait. As written above, some more commits from my side will follow. :smile:

**Pull request merged**